nat-lab/tests/conftest.py

```diff
@@ -192,69 +192,51 @@ async def perform_pretest_cleanups():
         await cleanup()
 
 
-async def _copy_binaries_on_ci(session):
-    mac_vm, win_vm_1, win_vm_2 = False, False, False
-
-    for item in session.items:
-        if "WINDOWS_VM_1" in item.name:
-            win_vm_1 = True
-
-        if "WINDOWS_VM_2" in item.name:
-            win_vm_2 = True
-
-        if "MAC_VM" in item.name:
-            mac_vm = True
-
-    if win_vm_1:
-        async with windows_vm_util.new_connection(
-            LAN_ADDR_MAP[ConnectionTag.WINDOWS_VM_1], copy_binaries=True
-        ):
-            pass
-
-    if win_vm_2:
-        async with windows_vm_util.new_connection(
-            LAN_ADDR_MAP[ConnectionTag.WINDOWS_VM_2], copy_binaries=True
-        ):
-            pass
-
-    if mac_vm:
-        async with mac_vm_util.new_connection(copy_binaries=True):
-            pass
+async def _copy_vm_binaries(tag: ConnectionTag):
+    is_ci = os.environ.get("CUSTOM_ENV_GITLAB_CI") is not None and os.environ.get(
+        "CUSTOM_ENV_GITLAB_CI"
+    )
+    if tag is ConnectionTag.WINDOWS_VM_1 or ConnectionTag.WINDOWS_VM_2:
+        try:
+            async with windows_vm_util.new_connection(
+                LAN_ADDR_MAP[tag], copy_binaries=True
+            ):
+                pass
+        except OSError as e:
+            if is_ci:
+                raise e
+            print(e)
+    elif tag is ConnectionTag.MAC_VM:
+        try:
+            async with mac_vm_util.new_connection(copy_binaries=True):
+                pass
+        except OSError as e:
+            if is_ci:
+                raise e
+            print(e)
 
 
-async def _copy_binaries_on_local_run():
-    try:
-        async with windows_vm_util.new_connection(
-            LAN_ADDR_MAP[ConnectionTag.WINDOWS_VM_1], copy_binaries=True
-        ):
-            pass
-    except OSError as e:
-        print(e)
+async def _copy_vm_binaries_if_needed(items):
+    windows_bins_copied = False
+    mac_bins_copied = False
 
-    try:
-        async with windows_vm_util.new_connection(
-            LAN_ADDR_MAP[ConnectionTag.WINDOWS_VM_2], copy_binaries=True
-        ):
-            pass
-    except OSError as e:
-        print(e)
+    for item in items:
+        for mark in item.own_markers:
+            if mark.name == "windows" and not windows_bins_copied:
+                await _copy_vm_binaries(ConnectionTag.WINDOWS_VM_1)
+                await _copy_vm_binaries(ConnectionTag.WINDOWS_VM_2)
+                windows_bins_copied = True
+            elif mark.name == "mac" and not mac_bins_copied:
+                await _copy_vm_binaries(ConnectionTag.MAC_VM)
+                mac_bins_copied = True
```

Comment:
Changed the logic so that we copy only based on the pytest.marks. (Also added a few marks to the tests that were missing it)